core/src/main/scala/com/evolutiongaming/catshelper/Log.scala

```diff
@@ -76,7 +78,92 @@ object Log {
 
   def summon[F[_]](implicit F: Log[F]): Log[F] = F
 
-  def apply[F[_]: Sync](logger: Logger): Log[F] = new Log[F] {
+  def cached[F[_] : Sync](source: String, factory: ILoggerFactory): Log[F] = new Log[F] {
```

Comment:
this is still an effectful unpure code, hence - **NO** from my side. however you can already use `logOf(…).toTry.get` in your code base, which will clearly express to readers unsafeness, and this is easy to use

Reply:
Why is that an effectual code?

Reply:
If you worry about getting the factory in safe manner then check out test file in this PR. It shows that we still should  get ILoggerFactory by delaying slf4j stuff.

Reply:
because any manipulations with `factory` are not safe, and in runtime it really references mutable shared state and change it

Reply:
Well, I see what you mean. It easily can be solved by putting function here (it should be done anyway because of Logback) like so
```scala
def cached[F[_] : Sync](source: String, factory: String => F[Logger]): Log[F] = ???
```.
Or by putting factory inside  of `Ref`. I guess it would be pure code then, right?

Reply:
Also I see that every manipulation with `factory` is delayed here, isn't it enough to be pure?

Reply:
right!

however such changes would enforce your `Log` implementation to call `factory.getLogger` as part of `log.info/debug/etc` - adds some runtime overhead

Reply:
Such runtime overhead is so little we can ignore it. If you dive in (not so deep) to the implementations, you can see that:
```scala
private Map<String, Logger> loggerCache;
//...
public final Logger getLogger(final String name) {

        if (name == null) {
            throw new IllegalArgumentException("name argument cannot be null");
        }

        // if we are asking for the root logger, then let us return it without
        // wasting time
        if (Logger.ROOT_LOGGER_NAME.equalsIgnoreCase(name)) {
            return root;
        }

        int i = 0;
        Logger logger = root;

        // check if the desired logger exists, if it does, return it
        // without further ado.
        Logger childLogger = (Logger) loggerCache.get(name);
        // if we have the child, then let us return it without wasting time
        if (childLogger != null) {
            return childLogger;
        }

        // if the desired logger does not exist, them create all the loggers
        // in between as well (if they don't already exist)

//....

}

```

Reply:
You know when you create instances of IO and flatMap classes for every operation you make this overhead seems totally fine.

Reply:
sure it might be totally fine, it depends on actual implementation :)

```
def cached[F[_] : Sync](source: String, factory: String => F[Logger]): Log[F]
```
signaturewise & typewise  - is ok